src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/StackItem.cs

```diff
@@ -3,68 +3,137 @@
+
+            return stackItem._int32;
         }
-    }
 
-    internal class Int64StackItem : StackItem<long>
-    {
-        public Int64StackItem(long value) : base(value, StackValueKind.Int64, WellKnownType.Int64)
+        public static implicit operator StackItem(long int64)
         {
+            return new StackItem { _int64 = int64, _type = WellKnownType.Int64, _kind = StackValueKind.Int64 };
         }
-    }
 
-    internal class FloatStackItem : StackItem<float>
-    {
-        public FloatStackItem(float value) : base(value, StackValueKind.Float, WellKnownType.Single)
+        public static explicit operator long(StackItem stackItem)
         {
+            if (stackItem._type != WellKnownType.Int64)
+                throw new InvalidCastException();
+
+            return stackItem._int64;
         }
-    }
 
-    internal class DoubleStackItem : StackItem<double>
-    {
-        public DoubleStackItem(double value) : base(value, StackValueKind.Float, WellKnownType.Double)
+        public static implicit operator StackItem(IntPtr ptr)
         {
+            return new StackItem { _nativeInt = ptr, _type = WellKnownType.IntPtr, _kind = StackValueKind.NativeInt };
         }
-    }
 
-    internal class ValueTypeStackItem : StackItem<ValueType>
-    {
-        public ValueTypeStackItem(ValueType value) : base(value, StackValueKind.ValueType, WellKnownType.ValueType)
+        public static explicit operator IntPtr(StackItem stackItem)
         {
+            if (stackItem._type != WellKnownType.IntPtr)
+                throw new InvalidCastException();
+
+            return stackItem._nativeInt;
         }
-    }
 
-    internal class ObjectRefStackItem : StackItem<Object>
-    {
-        public ObjectRefStackItem(Object value) : base(value, StackValueKind.ObjRef, WellKnownType.Object)
+        public static implicit operator StackItem(float single)
         {
+            return new StackItem { _float = single, _type = WellKnownType.Single, _kind = StackValueKind.Float };
+        }
+
+        public static explicit operator float(StackItem stackItem)
+        {
+            if (stackItem._type != WellKnownType.Single)
+                throw new InvalidCastException();
+
+            return stackItem._float;
+        }
+
+        public static implicit operator StackItem(double d)
+        {
+            return new StackItem { _double = d, _type = WellKnownType.Double, _kind = StackValueKind.Float };
+        }
+
+        public static explicit operator double(StackItem stackItem)
+        {
+            if (stackItem._type != WellKnownType.Double)
+                throw new InvalidCastException();
+
+            return stackItem._double;
+        }
+
+        public static StackItem FromValueType(ValueType valueType)
+        {
+            return new StackItem { _valueType = valueType, _type = WellKnownType.ValueType, _kind = StackValueKind.ValueType };
+        }
+
+        public ValueType ToValueType()
```

Comment:
`As*` may be a better name for these simple accessors, like `AsObjectRef()`, `AsDouble()`, ... . 

`To*` implies potential conversion. These simple accessors are not doing any conversion.